File: agents/java/src/main/java/com/nodetrace/services/RetryPolicy.java
package com.nodetrace.services;

import com.nodetrace.utils.Logger;

public class RetryPolicy {

    private final int maxAttempts;
    private final int baseDelayMs;

    public RetryPolicy(int maxAttempts, int baseDelayMs) {
        this.maxAttempts = maxAttempts;
        this.baseDelayMs = baseDelayMs;
    }

    public void execute(Runnable action) throws Exception {
        int attempt = 1;

        while (true) {
            try {
                action.run();
                return;
            } catch (Exception e) {
                if (attempt >= maxAttempts) {
                    Logger.error("RetryPolicy: max attempts reached.");
                    throw e;
                }

                int delay = baseDelayMs * attempt;
                Logger.warn("Retry attempt " + attempt + " failed. Retrying in " + delay + "ms...");
                Thread.sleep(delay);
                attempt++;
            }
        }
    }
}

File: agents/csharp/NodeTraceAgent/Utils/RetryPolicy.cs
namespace NodeTraceAgent.Utils
{
    public static class RetryPolicy
    {
        public static async Task<T?> ExecuteWithRetry<T>(
            Func<Task<T>> action,
            int maxRetries = 5,
            int initialDelayMs = 1000)
        {
            int attempt = 0;
            int delay = initialDelayMs;

            while (attempt < maxRetries)
            {
                try
                {
                    return await action();
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"[Retry] Attempt {attempt + 1}/{maxRetries} failed: {ex.Message}");
                }

                await Task.Delay(delay);
                delay *= 2; // backoff esponenziale
                attempt++;
            }

            return default;
        }
    }
}